python/lsst/meas/extensions/multiprofit/fit_coadd_multiband.py

```diff
@@ -0,0 +1,235 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from dataclasses import dataclass
+import math
+import numpy as np
+from typing import List, Sequence, Set
+
+import gauss2d as g2
+import gauss2d.fit as g2f
+
+import lsst.afw.image as afwImage
+from lsst.daf.butler.formatters.parquet import astropy_to_arrow
+import lsst.geom as geom
+import lsst.pex.config as pexConfig
+import lsst.pipe.base as pipeBase
+import lsst.pipe.tasks.fit_coadd_multiband as fitMB
+import lsst.utils.timer as utilsTimer
+
+from multiprofit.fit_psf import CatalogPsfFitterConfig
+from multiprofit.fit_source import CatalogExposureSourcesABC, CatalogSourceFitter, CatalogSourceFitterConfig
+
+from .utils import get_spanned_image
+
+
+class NotPrimaryError(RuntimeError):
+    pass
+
+
+@dataclass(frozen=True, kw_only=True, slots=True)
+class CatalogExposurePsfs(fitMB.CatalogExposureInputs, CatalogExposureSourcesABC):
+    """ Input data from lsst pipelines, parsed for MultiProFit.
+    """
+    channel: g2f.Channel
+    config_fit: CatalogSourceFitterConfig
+
+    def get_psfmodel(self, source):
+        match = np.argwhere(
+            (self.table_psf_fits[self.config_fit_psf.column_id] == source[self.config_fit.column_id])
+        )[0][0]
+        return self.config_fit_psf.rebuild_psfmodel(self.table_psf_fits[match])
+
+    def get_source_observation(self, source) -> g2f.Observation:
+        if not (source['detect_isPrimary'] and not source['merge_peak_sky']):
+            raise NotPrimaryError(f'source {source[self.config_fit.column_id]} has invalid flags for fit')
+        footprint = source.getFootprint()
+        bbox = footprint.getBBox()
+        spans = footprint.getSpans()
+        bitmask = 0
+        mask = self.exposure.mask.subset(bbox)
+        for bitname in self.config_fit.mask_names_zero:
+            bitval = mask.getPlaneBitMask(bitname)
+            bitmask |= bitval
+        mask = (mask.array & bitmask) != 0
+        mask_span = afwImage.Mask(bbox)
+        spans.setMask(mask_span, 2)
+        mask |= (mask_span.array == 0)
+        mask = ~mask
+        img, _, sigma_inv = get_spanned_image(footprint, bbox=bbox, spans=spans, get_sig_inv=True,
+                                              exposure=self.exposure)
+        sigma_inv[~mask] = 0
+
+        obs = g2f.Observation(image=g2.ImageD(img), sigma_inv=g2.ImageD(sigma_inv),
+                              mask_inv=g2.ImageB(mask), channel=self.channel)
+        return obs
+
+    def __post_init__(self):
+        # A truly hideous workaround to https://github.com/python/cpython/issues/83315
```

Comment:
I looked at your post in issue 83315 and I think it's due to you using the `ABCMeta` class as opposed to `ABC`. For example, this code works fine:

```
class A(ABC):
    @property
    @abstractmethod
    def x(self) -> int:
        return self.x

@dataclass(slots=True, kw_only=True)
class B(A):
    x: int
    
    def __post_init__(self):
        pass

@dataclass(slots=True, kw_only=True)
class C(B):
    y: int
        
    def __post_init_(self):
        super(B, self).__post_init__()
```
for both `B(x=1)` and `C(x=1, y=2)`.

Reply:
`class A(ABC)` and `class A(metaclass=ABCMeta)` do the same thing (although fair enough that the former is newer and more readable, so it should be used in new code). The example you posted only works because you're missing the second underscore at the end of `C.__post_init_` so it's not actually being called.

Reply:
Yup, you're right. That is weird too. I'm not sure what the appropriate syntax should be. I almost feel like if `B` is a a `dataclass`, and `C` inherits from `B`, then `C` should be a `dataclass` and you should just be able to add attributes. But apparently that isn't the case.